feature/amazon/src/main/java/com/revenuecat/purchases/amazon/handler/PurchaseHandler.kt

```diff
@@ -0,0 +1,120 @@
+        }
+
+        val requestId = PurchasingService.purchase(productDetails.sku)
+        purchaseCallbacks[requestId] = onSuccess to onError
+    }
+
+    override fun onPurchaseResponse(response: PurchaseResponse) {
+        log(LogIntent.DEBUG, AmazonStrings.PURCHASE_REQUEST_FINISHED.format(response.toJSON().toString(1)))
+
+        val requestId = response.requestId
+        val callbacks = purchaseCallbacks.remove(requestId)
+
+        callbacks?.let { (onSuccess, onError) ->
+            when (response.requestStatus) {
+                PurchaseResponse.RequestStatus.SUCCESSFUL ->
+                    onSuccessfulPurchase(response.receipt, response.userData, onSuccess)
+                PurchaseResponse.RequestStatus.FAILED -> onFailed(onError)
+                PurchaseResponse.RequestStatus.INVALID_SKU -> onInvalidSku(onError)
+                PurchaseResponse.RequestStatus.ALREADY_PURCHASED -> onAlreadyPurchased(onError)
+                PurchaseResponse.RequestStatus.NOT_SUPPORTED -> onNotSupported(onError)
+                null -> onUnknownError(onError)
+            }
+        }
+    }
+
+    @SuppressWarnings("ForbiddenComment")
+    private fun onSuccessfulPurchase(
+        receipt: Receipt,
+        userData: UserData,
+        onSuccess: (Receipt, UserData) -> Unit
+    ) {
+        if (receipt.isCanceled) {
+            // TODO: Do we do anything?
+        } else {
+            onSuccess(receipt, userData)
+        }
+    }
+
+    private fun onUnknownError(onError: (PurchasesError) -> Unit) {
+        PurchasesError(
+            PurchasesErrorCode.StoreProblemError,
+            "Failed to make purchase. There was an Amazon store problem"
+        ).let {
+            onError(it)
+        }
+    }
+
+    private fun onNotSupported(onError: (PurchasesError) -> Unit) {
+        PurchasesError(
+            PurchasesErrorCode.StoreProblemError,
+            "Failed to make purchase. Call is not supported"
+        ).let {
+            onError(it)
+        }
+    }
+
+    private fun onAlreadyPurchased(onError: (PurchasesError) -> Unit) {
+        PurchasesError(
+            PurchasesErrorCode.ProductAlreadyPurchasedError,
+            "Failed to make purchase. Product already owns SKU."
+        ).let {
+            onError(it)
+        }
+    }
+
+    private fun onInvalidSku(onError: (PurchasesError) -> Unit) {
+        PurchasesError(
+            PurchasesErrorCode.ProductNotAvailableForPurchaseError,
+            "Failed to make purchase. SKU is invalid"
+        ).let {
+            onError(it)
+        }
+    }
+
+    private fun onFailed(onError: (PurchasesError) -> Unit) {
+        // Indicates that the purchase failed. Can simply mean user cancelled
+        PurchasesError(
+            PurchasesErrorCode.PurchaseCancelledError,
+            "Failed to make purchase. This error normally means that the purchase was cancelled"
```

Comment:
That's a good point. There's no deferred billing (https://developer.amazon.com/docs/in-app-purchasing/iap-migrate-from-google-iab-to-amazon-iap.html#feature-comparison) but it's possible to trigger a EU MFA flow, which in iOS behaves similarly to deferred, but not in Amazon.

During a EU MFA flow, it will actually send `SUCCESSFUL` when completing the MFA, and failed if there was something wrong during the challenge phase. 

>  If you want to make sure your apps’ implementation of in-app purchase is ready for MFA, you just need to verify that the app correctly follows the steps described in our in-app purchase documentation on the Amazon Developers portal.

> The multi-factor authentication system used by Amazon follows the “pessimistic fulfilment” approach.

> The “pessimistic fulfilment” approach means that if the customer correctly completes all the steps necessary to satisfy the multi-factor authentication, then the transaction will be correctly processed as successful. In regards of the Amazon in-app purchase implementation, this means the method response.getRequestStatus() in onPurchaseResponse() will return case SUCCESSFUL. At this stage the in-app item can then be fulfilled to the customer.

> However, if anything happens during the “challenge phase” of the MFA process (when the customer is prompted to go on their bank’s multi-factor authentication system), for example the customer clicks the “back button” or exits the window showing the MFA challenge, then the response.getRequestStatus() method will return case FAILED, and your app will need to gracefully manage the failed transaction.

> Note: There’s one exception to pessimistic fulfillment, and that is for subscriptions with free trial: when customers purchase a subscription with a free trial, the purchase will be fulfilled optimistically, without presenting an MFA challenge. The MFA challenge will be then mandated when free trial expires.

https://developer.amazon.com/blogs/appstore/post/72ee4001-9e54-49b1-83ad-f1cede9b40ce/ensuring-your-implementation-of-in-app-purchase-is-ready-for-eu-multi-factor-authentication



I haven't tested this scenario btw and I am trusting the documentation. Do you think we should change the messaging of this error to something more generic? I am not that comfortable about passing the purchase as cancelled, but there's no other way to find out if the purchase was cancelled.

Reply:
🤔 maybe we can add a comment and point to docs. we can probably set up a `rev.cat` link to amazon's docs for now and then swap them for our own later.